### ISSUE TITLE
Complicated TPT model, throws error once relationships are created > 1 layer deep

### ISSUE THREAD

**Issue by jdlegan** [R4ND3LL/EdmxRuler#27]
R4ND3LL,

I am attempting to migrate a largeish db model which is TPT heavy. Many years ago I forked the code for EF power tools and wrote a lot of T4 template code to handle all of the inheritance generation, removing navigation properties and properties implemented in base classes, up to three layers deep. I was about to do the same now that the project is moving to EF Core 7 from EF6 when I saw your conversation with Erik regarding your tool and how it worked. I spent a few hours trying out various features and got the model where I would like it and then started adding in support for base types after my initial testing showed it worked flawlessly. 

As soon as I started getting into the more complicated TPT scenarios (example below), I started getting the following error: 

`A key cannot be configured on 'XXXXXX (Dictionary<string, object>)' because it is a derived type. The key must be configured on the root type 'YYYYYYY (Dictionary<string, object>)'. If you did not intend for 'YYYYYYY (Dictionary<string, object>)' to be included in the model, ensure that it is not referenced by a DbSet property on your context, referenced in a configuration call to ModelBuilder, or referenced from a navigation on a type that is included in the model.`

The scenario in question is one used throughout the database:

Table 1.0 w/discriminator

Table 1.1, 1.2, 1.3 w/FK to and computed persisted column of discriminator

For each of Table 1.1 - 1.3, potentially further breakdown (i.e. Table 1.1.1, Table 1.1.2 and even Table 1.1.2.1)

The model works fine (ignoring performance for a moment) and has worked fine in EF6 for over six years. I tried to replicate the existing EF6 model, which does not have navigation properties on any of the TPT tables through their PK other than at the root Table 1.0 level except for those that are unique to it by marking them as "Not mapped" in the GUI and re-running the scaffold but so far to no avail. I am going to grab the source on Monday to debug but wanted to shoot this message over in case there was something fundamental that I am missing as I am new to EF Core but have extensive experience in EF6. 

It is worth noting that there is no EDMX at play here, this is just the ef scaffold command running.

**Comment by R4ND3LL** [comment]
In TPT, the base entity should have its primary key mapped normally, but for each derived type, the corresponding primary key should be checked "not mapped" in the EFR rule editor.  (or in the json file: "NotMapped": true ).   That seems to be specifically what the error is complaining about.

To streamline this configuration, I may force the killing of primary keys on derived types, as well as the FKs that usually exist between the tables in the hierarchy.  

If you can mock up a small database table scenario that replicates the error, I'd be happy to script it in and test it.

**Comment by jdlegan** [comment]
Thank you for the offer. I am going to give it another shot tomorrow. The one thing I noticed is that "IsKey" is not checked for any tables, TPT or otherwise, if I create a new json file by running scaffold with the Design package in the project. Is this expected behavior?

**Comment by jdlegan** [comment]
I am seeing the same behavior with the sample database I created today.  Fairly simple TPT, removed composite keys for enforcement of derived tables w/persisted computed keys. Vehicles -> Cars | Trucks -> Sports Cars or Family Cars | Light Duty Trucks or Heavy Duty Trucks. 

[TPTRepro.zip](https://github.com/R4ND3LL/EntityFrameworkRuler/files/12009003/TPTRepro.zip)

**Comment by jdlegan** [comment]
I looked at the source for a few minutes last night and something jumped out at me so on a hunch I gave it a quick test while on a call. If I do not reuse the primary key name (VehicleId) in all derived tables, levels 2-N (the example database is TPT three levels deep), other than needing to flag the keys as not mapped, it produces the expected classes, each extending its base has as expected (i.e. LightDutyTruck -> Truck -> Vehicle). However, in my experience, most TPT models preserve the naming convention of the based table's PK throughout the derived table meaning VehicleId is present in all layers. The change I made was to rename the columns in all derived tables to match the concrete table name (for Light Duty Truck the PK was renamed from VehicleId to LightDutyTruckVehicleId and the PK in Truck was renamed from VehicleId to TruckVehicleId). This might be at the root of everything I am seeing on the real database as the base tables PK name is persisted throughout every level of the model.

**Comment by R4ND3LL** [comment]
Thanks for the sample.  I reproduced and fixed the issue on build v1.2.33.

I setup the TPT hierarchy in the rules file (see attached) for the truck side only.

Nothing major you have to configure. Just set the following annotation for Vehicle:

Relational:MappingStrategy = TPT

And ensure the derived types have "Vehicle" as the base type, or what ever is appropriate for deeper levels..  

When scaffolding, EFR will now automatically ignore FK relations within the TPT hierarchy, which is what is causing the error.

Vehicle data loaded as expected during my testing.

Let me know if you have any further issues.

[Vehicles-rules.zip](https://github.com/R4ND3LL/EntityFrameworkRuler/files/12022427/Vehicles-rules.zip)

**Comment by jdlegan** [comment]
@R4ND3LL ,

It appears to have solved the initially reported condition, thank you! I am going to grab the previous version because I am fairly certain I am seeing regression on another, though I might be confusing the results with the tests I was performing with Erik's tool. 

Using your configuration file, without any changes, I see the following class being generated for `Trucks`:

```
namespace TPT_Repro.Models;

public partial class Truck : Vehicle 
{
    public int TruckTypeId { get; set; }

    public string TruckNote { get; set; } = null!;

    public virtual TruckType TruckType1 { get; set; } = null!;
}
```

Which correctly extends Vehicle and omits the PK `VehicleId` as desired. However the name of the navigation property `TruckType1` is caused by the relationship between `Vehicle` and `TruckType`. The way we have historically solved this is by setting a namespace at the various TPT levels so instead of `namespace TPT_Repro.Models;`, this would be `namespace TPT_Repro.Models.Vehicles.Trucks;`. I am fairly confident that changing the namespace was working in the previous release but setting the namespace on the schema is having no impact on the generated classes. The json for `Trucks` is as follows

```
"Schemas": [
    {
      "SchemaName": "Trucks",
      "IncludeUnknownTables": true,
      "IncludeUnknownViews": true,
      "IncludeUnknownFunctions": false,
      "TableRegexPattern": "^(.*)$",
      "TablePatternReplaceWith": "$1Entity",
      "Namespace": "TPT_Repro.Models.Vehicles.Trucks",
      "Entities": [...]
      ...
```

 but the schema remains `TPT_Repro.Models` on all generated classes. Is this regression or am I misremembering it working?

**Comment by R4ND3LL** [comment]
the Namespace property has never had any effect on scaffolded entities.  Perhaps EF Power Tools has a similar property and does something with it.  EFR extends on EF Core's native scaffolding features by plugging into their design-time services.  As such, EF Core only takes a single namespace argument up front and uses that when outputting all entity files.  This area is fairly closed off, from an extension/override perspective.  

That said, there should be no real reason the nav prop cant be generated as  "TruckType".  I'll start a new issue for that and have a look.

**Comment by R4ND3LL** [comment]
Issue of entities involved in inheritance receiving unexpected navigation property names has been resolved in EFR Design package v1.2.36
